benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,562 @@
+                self.futures[bp_loc] = [fut]
+            else:
+                if self.batch_mode:
+                    self.gradient[bp_loc].append(gradient)
+                else:
+                    self.gradient[bp_loc] += gradient
+                self.futures[bp_loc].append(fut)
+            if len(self.futures[bp_loc]) == self.trainer_count:
+                if self.batch_mode:
+                    # TODO: cuda kernel
+                    bp_loc_avg = self.gradient[bp_loc][0]
+                    for i in range(1, self.trainer_count):
+                        bp_loc_avg += self.gradient[bp_loc][i]
+                else:
+                    bp_loc_avg = self.gradient[bp_loc]
+                bp_loc_avg / (1.0 * self.trainer_count)
+                if not self.use_cuda_rpc:
+                    bp_loc_avg = bp_loc_avg.cpu()
+                if sparse_gradient:
+                    if self.backend == GLOO:
+                        bp_loc_avg = bp_loc_avg.coalesce()
+                    bp_loc_avg = [bp_loc_avg._indices(), bp_loc_avg._values(), bp_loc_avg.size()]
+                for cur_fut in self.futures[bp_loc]:
+                    cur_fut.set_result(bp_loc_avg)
+                return fut
+        return fut
+
+    def clear_state(self):
+        self.futures.clear()
+        self.gradient.clear()
+
+    def reset(gs_rref):
+        self = gs_rref.local_value()
+        self.futures.clear()
+        self.gradient.clear()
+        self.batch_number = 0
+
+# --------------------------- Model -----------------------------------------
+
+
+class DummyModel(nn.Module):
+    def __init__(self, num_embeddings=4, embedding_dim=4, dense_input_size=4, dense_output_size=4, sparse=True):
+        super().__init__()
+        self.embedding = nn.EmbeddingBag(num_embeddings, embedding_dim, sparse=sparse)
+        self.fc1 = nn.Linear(dense_input_size, dense_output_size)
+
+    def forward(self, x):
+        x = self.embedding(x)
+        return F.softmax(self.fc1(x), dim=1)
+
+def get_model(model_id, model_config):
+    if model_id == 1:
+        return DummyModel(**model_config)
+    sys.exit("model_id not found")
+
+# --------------------------- Data -----------------------------------------
+
+
+class RandomData:
+    def __init__(self, min_val: int = 0, max_val: int = 4, batch_size: int = 4, mult: int = 2):
+        self.input = torch.randint(min_val, max_val, [batch_size, mult])
+        self.target = torch.randint(min_val, max_val, [batch_size])
+
+    def get_input_and_target(self):
+        return self.input, self.target
+
+def get_data(data_id, data_config):
+    if data_id == 1:
+        return RandomData(**data_config)
+    sys.exit("data_id not found")
+
+# --------------------------- Hooks -----------------------------------------
+
+#TODO look into refactoring this method
+def register_hook(hook_id, ddp_model, process_group, gs_rref, world_size, rank, use_cuda_rpc, futures, bp_location, batch_number, metrics):
```

Comment:
+1 we need to refactor this.

Reply:
Created a HookState class and passed it as the state parameter when registering the communication hook.